mediagrains_py36/numpy/__init__.py

```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/python
+#
+# Copyright 2019 British Broadcasting Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""\
+Library for handling mediagrains in numpy arrays
+"""
+
+from mediagrains.cogenums import (
+    CogFrameFormat,
+    COG_FRAME_IS_PACKED,
+    COG_FRAME_IS_COMPRESSED,
+    COG_FRAME_FORMAT_BYTES_PER_VALUE)
+from mediagrains import grain as bytesgrain
+from mediagrains import grain_constructors as bytesgrain_constructors
+from copy import copy, deepcopy
+
+import numpy as np
+from numpy.lib.stride_tricks import as_strided
+
+
+__all__ = ['VideoGrain', 'VIDEOGRAIN']
+
+
+def _dtype_from_cogframeformat(fmt: CogFrameFormat) -> np.dtype:
+    if not COG_FRAME_IS_PACKED(fmt) and not COG_FRAME_IS_COMPRESSED(fmt):
+        if COG_FRAME_FORMAT_BYTES_PER_VALUE(fmt) == 1:
+            return np.dtype(np.uint8)
+        elif COG_FRAME_FORMAT_BYTES_PER_VALUE(fmt) == 2:
+            return np.dtype(np.int16)
+        elif COG_FRAME_FORMAT_BYTES_PER_VALUE(fmt) == 4:
+            return np.dtype(np.int32)
+    elif fmt in [CogFrameFormat.UYVY,
+                 CogFrameFormat.YUYV,
+                 CogFrameFormat.RGB,
+                 CogFrameFormat.RGBx,
+                 CogFrameFormat.RGBA,
+                 CogFrameFormat.BGRx,
+                 CogFrameFormat.BGRx,
+                 CogFrameFormat.ARGB,
+                 CogFrameFormat.xRGB,
+                 CogFrameFormat.ABGR,
+                 CogFrameFormat.xBGR]:
+        return np.dtype(np.uint8)
+    elif fmt == CogFrameFormat.v216:
+        return np.dtype(np.int16)
+    elif fmt == CogFrameFormat.v210:
+        return np.dtype(np.int32)
+
+    raise NotImplementedError("Cog Frame Format not amongst those supported for numpy array interpretation")
+
+
+def _component_arrays_for_data_and_type(data: np.ndarray, fmt: CogFrameFormat, components: bytesgrain.VIDEOGRAIN.COMPONENT_LIST):
+    if not COG_FRAME_IS_PACKED(fmt) and not COG_FRAME_IS_COMPRESSED(fmt):
+        arrays = []
+        for component in components:
+            component_data = data[component.offset//data.itemsize:(component.offset + component.length)//data.itemsize]
```

Comment:
Huh, well the floor quotient operator is a bit of syntax I never knew...

Reply:
It was rarely used in Python 2, but becomes more important in Py3, since `int/int=float` but `int//int=int`.